### ISSUE TITLE
Support QuerySets/lists of objects

### ISSUE THREAD

**Issue by rixx** [raphaelm/django-hierarkey#10]
If I want to save a queryset, atm I do this:

```
hierarkey.add_type(
    QuerySet,
    lambda queryset: ','.join([str(element.pk) for element in queryset]),
    lambda pk_list: [MyModel.objects.get(pk=element) for element in pk_list.split(',')]
)
```

– which works because I only ever use one type here. Sure, it's not a complete roundtrip (QuerySet to list of instances), but both are iterables of elements, which is good enough for my purposes. I understand if you don't want to make this behaviour default, but you might add a helper that one could then explicitly enable/add as type, so that nobody else needs to go through this ;).

Since the QuerySet knows about its model, there shouldn't be any need for hardcoding at all :)

**Comment by rixx** [comment]
Also, while I'm at it: the documentation regarding `add_type` was fairly unclear to me and I had to dig around quite a bit to find out where to find it. I'm not really sure what the preferred way is, though, so I didn't submit a doc PR for that.

**Comment by raphaelm** [comment]
> Since the QuerySet knows about its model, there shouldn't be any need for hardcoding at all :)

I'm not quite sure how this should work for unserializing if we just store lists of primary  keys, how would you know the model?

OTOH, hierarkey can already store models as ``{"model": "Name", "pk": 13}`` and can store lists, so we'd just need to put these two together.

> Also, while I'm at it: the documentation regarding add_type was fairly unclear to me and I had to dig around quite a bit to find out where to find it.

Do you refer to https://django-hierarkey.readthedocs.io/en/latest/exttype.html ? I think the example is simple enough, but yes, it could probably use some explanations.

**Comment by rixx** [comment]
> OTOH, hierarkey can already store models as {"model": "Name", "pk": 13} and can store lists, so we'd just need to put these two together.

That's exactly what I was thinking about – I just use the simplified version 'cause I'm lazy and know which models to expect.

> Do you refer to https://django-hierarkey.readthedocs.io/en/latest/exttype.html ? I think the example is simple enough

That's exactly what I was referring to, and imo the example is *too* minimal – as I said, I had to dig around in the library code to find out where to get the `hierarkey` variable from, that is magically there in the example.

**Comment by raphaelm** [comment]
We just closed #12 after an internal discussion because of two reasons:

* Loading model classes too dynamically based on serialized data is a remote code execution waiting to happen. With our current model serialization, we don't have that problem, as you explicitly need to specify the class to be loaded via ``as_type=ModelClass`` during de-serialization. We can't do ``as_type=QuerySet``, as this would require dynamic loading of the model class, we'd need some version of ``as_type=ContainerType[ModelClass]`` which doesn't feel very pythonic to me – in the end, probably less readable/predictable than storing lists in hierarkey and (de)serializing manually every time you use it with ``foo.settings.bar = [e.pk for e in queryset]`` and ``ModelClass.objects.filter(id__in=foo.settings.get(bar, as_type=list))``

* I believe hierarkey and the rest of your database should be viewed as two separate data stores and referencing models from hierarkey values should strongly be discouraged. It doesn't allow for foreign key constraint checks, reverse lookups, database joins. Hierarkey settings are meant to be a key-value store that makes sense in itself and can e.g. easily be copied over to a different system or to a different user in the same system. Storing references to models inside hierarkey breaks these assumptions and requires very special care when doing any of these operations and is therefore not something I want to officially support here. I know it sometimes is worth to use it as a hack, but that should be up to the user of the library to do. In pretix core, we do it at exactly one place, and it caused multiple problems over the years.
